src/t2c_template.c

```diff
@@ -0,0 +1,847 @@
+            T2C_LLVM_GEN_STORE_IMM32(builder3, ir->pc + 4, addr_PC);        \
+            LLVMBuildRetVoid(builder3);                                     \
+        }                                                                   \
+        LLVMBuildCondBr(*builder, cmp, taken, untaken);                     \
+    })
+
+BRANCH_FUNC(beq, EQ)
+BRANCH_FUNC(bne, NE)
+BRANCH_FUNC(blt, SLT)
+BRANCH_FUNC(bge, SGE)
+BRANCH_FUNC(bltu, ULT)
+BRANCH_FUNC(bgeu, UGE)
+
+T2C_OP(lb, {
+    LLVMValueRef mem_loc = t2c_gen_mem_loc(start, builder, ir, mem_base);
+    LLVMValueRef res = LLVMBuildSExt(
+        *builder, LLVMBuildLoad2(*builder, LLVMInt8Type(), mem_loc, "res"),
+        LLVMInt32Type(), "sext8to32");
+    LLVMBuildStore(*builder, res, t2c_gen_rd_addr(start, builder, ir));
+})
+
+T2C_OP(lh, {
+    LLVMValueRef mem_loc = t2c_gen_mem_loc(start, builder, ir, mem_base);
+    LLVMValueRef res = LLVMBuildSExt(
+        *builder, LLVMBuildLoad2(*builder, LLVMInt16Type(), mem_loc, "res"),
+        LLVMInt32Type(), "sext16to32");
+    LLVMBuildStore(*builder, res, t2c_gen_rd_addr(start, builder, ir));
+})
+
+
+T2C_OP(lw, {
+    LLVMValueRef mem_loc = t2c_gen_mem_loc(start, builder, ir, mem_base);
+    LLVMValueRef res =
+        LLVMBuildLoad2(*builder, LLVMInt32Type(), mem_loc, "res");
+    LLVMBuildStore(*builder, res, t2c_gen_rd_addr(start, builder, ir));
+})
+
+T2C_OP(lbu, {
+    LLVMValueRef mem_loc = t2c_gen_mem_loc(start, builder, ir, mem_base);
+    LLVMValueRef res = LLVMBuildZExt(
+        *builder, LLVMBuildLoad2(*builder, LLVMInt8Type(), mem_loc, "res"),
+        LLVMInt32Type(), "zext8to32");
+    LLVMBuildStore(*builder, res, t2c_gen_rd_addr(start, builder, ir));
+})
+
+T2C_OP(lhu, {
+    LLVMValueRef mem_loc = t2c_gen_mem_loc(start, builder, ir, mem_base);
+    LLVMValueRef res = LLVMBuildZExt(
+        *builder, LLVMBuildLoad2(*builder, LLVMInt16Type(), mem_loc, "res"),
+        LLVMInt32Type(), "zext16to32");
+    LLVMBuildStore(*builder, res, t2c_gen_rd_addr(start, builder, ir));
+})
+
+T2C_OP(sb, {
+    LLVMValueRef mem_loc = t2c_gen_mem_loc(start, builder, ir, mem_base);
+    T2C_LLVM_GEN_LOAD_VMREG(rs2, 8, t2c_gen_rs2_addr(start, builder, ir));
+    LLVMBuildStore(*builder, val_rs2, mem_loc);
+})
+
+T2C_OP(sh, {
+    LLVMValueRef mem_loc = t2c_gen_mem_loc(start, builder, ir, mem_base);
+    T2C_LLVM_GEN_LOAD_VMREG(rs2, 16, t2c_gen_rs2_addr(start, builder, ir));
+    LLVMBuildStore(*builder, val_rs2, mem_loc);
+})
+
+T2C_OP(sw, {
+    LLVMValueRef mem_loc = t2c_gen_mem_loc(start, builder, ir, mem_base);
+    T2C_LLVM_GEN_LOAD_VMREG(rs2, 32, t2c_gen_rs2_addr(start, builder, ir));
+    LLVMBuildStore(*builder, val_rs2, mem_loc);
+})
+
+T2C_OP(addi, {
+    T2C_LLVM_GEN_LOAD_VMREG(rs1, 32, t2c_gen_rs1_addr(start, builder, ir));
+    LLVMValueRef res = T2C_LLVM_GEN_ALU32_IMM(Add, val_rs1, ir->imm);
+    LLVMBuildStore(*builder, res, t2c_gen_rd_addr(start, builder, ir));
```

Comment:
Can we avoid the storing of result after computing here?

Reply:
Yes, it will be optimized through LLVM Pass.